contracts/strategies/traderjoe/TraderJoe.sol

```diff
@@ -84,62 +171,76 @@ contract TraderJoe is UUPSUpgradeable, StrategyOwnablePausableBaseUpgradeable {
+            amountX = depositTokenAmount;
+            amountY = pairDepositTokenAmount;
+        } else {
+            amountX = pairDepositTokenAmount;
+            amountY = depositTokenAmount;
+        }
 
-        // swapping back extra pairDepositToken to depositToken
-        uint256 extraPairDepositTokenAmount = strategyStorage
-            .pairDepositToken
-            .balanceOf(address(this)) - pairDepositTokenBalanceBeforeSwap;
-        if (extraPairDepositTokenAmount != 0) {
-            path[0] = address(strategyStorage.pairDepositToken);
-            path[1] = address(depositToken);
-            SwapServiceLib.swapExactTokensForTokens(
-                swapService,
-                extraPairDepositTokenAmount,
+        // Prepare params.
+        uint256 binsAmount = strategyStorage.binIds.length;
+        uint256[] memory distributionX = new uint256[](binsAmount);
+        uint256[] memory distributionY = new uint256[](binsAmount);
+
+        for (uint256 i; i < binsAmount; ++i) {
+            // Bin allocation has precision of 1e3 and TraderJoe V2 has 1e18.
+            distributionX[i] = strategyStorage.binAllocationsX[i] * 1e15;
+            distributionY[i] = strategyStorage.binAllocationsY[i] * 1e15;
+        }
+
+        int256[] memory deltaIds = new int256[](binsAmount);
+        (, , uint256 activeId) = strategyStorage.lbPair.getReservesAndId();
+
+        for (uint256 i; i < binsAmount; ++i) {
+            deltaIds[i] = int256(strategyStorage.binIds[i]) - int256(activeId);
+        }
+
+        // Deposit.
+        ITraderJoeLBRouter.LiquidityParameters memory liquidityParameters = ITraderJoeLBRouter
+            .LiquidityParameters(
+                address(strategyStorage.tokenX),
+                address(strategyStorage.tokenY),
+                strategyStorage.binStep,
+                amountX,
+                amountY,
+                0, // Base contracts take care of min amount.
+                0, // Base contracts take care of min amount.
+                activeId,
                 0,
-                path
+                deltaIds,
+                distributionX,
+                distributionY,
+                address(this),
+                // solhint-disable-next-line not-rely-on-time
+                block.timestamp
             );
-        }
 
-        // staking LP tokens on Trader Joe
-        strategyStorage.lpToken.approve(
-            address(strategyStorage.masterChef),
-            lpBalance
+        strategyStorage.tokenX.approve(
+            address(strategyStorage.lbRouter),
+            amountX
         );
-        strategyStorage.masterChef.deposit(strategyStorage.farmId, lpBalance);
+        strategyStorage.tokenY.approve(
+            address(strategyStorage.lbRouter),
+            amountY
+        );
+
+        strategyStorage.lbRouter.addLiquidity(liquidityParameters);
+
+        // Since the amount of remaining pairDepositToken is nearly zero, skip swapping it back to depositToken.
```

Comment:
I think we should at least provide some asserts to make sure that is true. Otherwise our users can lose money. What do you think?

Reply:
In V2, the amount of refund token comes from precision error during rounding down. That's why it's around 1~2. (not $1 but $1e-6)

I added calling getAmounts to see if it's amount is large enough that we can swap to USDC and swap it only if it's available but removed it because it looks like unnecessary gas cost.

Reply:
on sidechains, it is okay to use a bit more gas, the main thing is correctness. I am more concerned of things like this: https://github.com/block42-blockchain-company/bro-strategies/pull/34/files/d44e48cccd94bfed7a9f1f0c708d0797c977fe3b#r1044096386

Can we add the asserts back please?

Reply:
Avalanche is not sidechain.

Reply:
Will add something.